lib/std/Build/Step/Run.zig

```diff
@@ -588,12 +584,81 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
         argv_list.items[placeholder.index] = cli_arg;
     }
 
-    try runCommand(self, argv_list.items, has_side_effects, &digest, prog_node);
+    try runCommand(self, argv_list.items, has_side_effects, tmp_dir_path, prog_node);
 
     if (self.dep_output_file) |dep_output_file|
         try man.addDepFilePost(std.fs.cwd(), dep_output_file.generated_file.getPath());
 
+    const digest = man.final();
+
+    const any_output = output_placeholders.items.len > 0 or
+        self.captured_stdout != null or self.captured_stderr != null;
+
+    // Rename into place
+    if (any_output) {
+        const o_sub_path = "o" ++ fs.path.sep_str ++ &digest;
+        b.cache_root.handle.makePath(o_sub_path) catch |err| {
+            return step.fail("unable to make path '{}{s}': {s}", .{
+                b.cache_root, o_sub_path, @errorName(err),
+            });
+        };
```

Comment:
I don't understand how this PR is solving your problem because if you create the directory like this then it should always cause the next line to fail when it tries to overwrite a directory with `rename`.

Instead, it should optimistically try the rename first, and then create *only the 'o' directory* if it needs to based on the error code.